File: app/src/main/java/de/westnordost/streetcomplete/data/osm/persist/ElementGeometryDao.java
package de.westnordost.streetcomplete.data.osm.persist;


import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteStatement;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import javax.inject.Inject;

import de.westnordost.streetcomplete.data.osm.ElementGeometry;
import de.westnordost.streetcomplete.util.Serializer;
import de.westnordost.osmapi.map.data.Element;
import de.westnordost.osmapi.map.data.LatLon;
import de.westnordost.osmapi.map.data.OsmLatLon;

public class ElementGeometryDao
{
	private final SQLiteOpenHelper dbHelper;
	private final Serializer serializer;

	private final SQLiteStatement insert;

	public static class Row
	{
		public Row(Element.Type elementType, long elementId, ElementGeometry geometry)
		{
			this.elementType = elementType;
			this.elementId = elementId;
			this.geometry = geometry;
		}

		public Element.Type elementType;
		public long elementId;
		public ElementGeometry geometry;
	}

	@Inject public ElementGeometryDao(SQLiteOpenHelper dbHelper, Serializer serializer)
	{
		this.dbHelper = dbHelper;
		this.serializer = serializer;

		String sql = "INSERT OR REPLACE INTO " + ElementGeometryTable.NAME + " ("+
				ElementGeometryTable.Columns.ELEMENT_TYPE+","+
				ElementGeometryTable.Columns.ELEMENT_ID+","+
				ElementGeometryTable.Columns.GEOMETRY_POLYGONS+","+
				ElementGeometryTable.Columns.GEOMETRY_POLYLINES+","+
				ElementGeometryTable.Columns.LATITUDE+","+
				ElementGeometryTable.Columns.LONGITUDE+
				") values (?,?,?,?,?,?);";
		SQLiteDatabase db = dbHelper.getWritableDatabase();
		insert = db.compileStatement(sql);
	}

	public void putAll(Collection<Row> rows)
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();

		db.beginTransaction();

		for(Row row : rows)
		{
			executeInsert(row.elementType, row.elementId, row.geometry);
		}

		db.setTransactionSuccessful();
		db.endTransaction();
	}

	/** adds or updates (overwrites) an element geometry*/
	public void put(Element.Type type, long id, ElementGeometry geometry)
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();
		db.beginTransaction();
		executeInsert(type, id, geometry);
		db.setTransactionSuccessful();
		db.endTransaction();
	}

	private void executeInsert(Element.Type type, long id, ElementGeometry geometry)
	{
		insert.bindString(1, type.name());
		insert.bindLong(2, id);
		if (geometry.polygons != null)
			insert.bindBlob(3, serializer.toBytes(geometry.polygons));
		else
			insert.bindNull(3);
		if (geometry.polylines != null)
			insert.bindBlob(4, serializer.toBytes(geometry.polylines));
		else
			insert.bindNull(4);
		insert.bindDouble(5, geometry.center.getLatitude());
		insert.bindDouble(6, geometry.center.getLongitude());

		insert.executeInsert();
		insert.clearBindings();
	}

	public ElementGeometry get(Element.Type type, long id)
	{
		SQLiteDatabase db = dbHelper.getReadableDatabase();
		String where = ElementGeometryTable.Columns.ELEMENT_TYPE + " = ? AND " +
				ElementGeometryTable.Columns.ELEMENT_ID + " = ?";
		String[] args = {type.name(), String.valueOf(id)};

		try (Cursor cursor = db.query(ElementGeometryTable.NAME, null, where, args,
				null, null, null, "1"))
		{
			if (!cursor.moveToFirst()) return null;
			return createObjectFrom(serializer, cursor);
		}
	}

	static ElementGeometry createObjectFrom(Serializer serializer, Cursor cursor)
	{
		int colGeometryPolygons = cursor.getColumnIndexOrThrow(ElementGeometryTable.Columns.GEOMETRY_POLYGONS),
			colGeometryPolylines = cursor.getColumnIndexOrThrow(ElementGeometryTable.Columns.GEOMETRY_POLYLINES),
			colCenterLat = cursor.getColumnIndexOrThrow(ElementGeometryTable.Columns.LATITUDE),
			colCenterLon = cursor.getColumnIndexOrThrow(ElementGeometryTable.Columns.LONGITUDE);

		List<List<LatLon>> polygons = null, polylines = null;

		if(!cursor.isNull(colGeometryPolygons))
		{
			polygons = serializer.toObject(cursor.getBlob(colGeometryPolygons), ArrayList.class);
		}
		if(!cursor.isNull(colGeometryPolylines))
		{
			polylines = serializer.toObject(cursor.getBlob(colGeometryPolylines), ArrayList.class);
		}
		LatLon center = new OsmLatLon(cursor.getDouble(colCenterLat), cursor.getDouble(colCenterLon));
		return new ElementGeometry(polylines, polygons, center);
	}

	/** Cleans up element geometry entries that belong to elements that are not referenced by any
	 *  quest anymore. */
	public int deleteUnreferenced()
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();

		/* SQLite does not allow selecting multiple columns in a DELETE subquery. Using a workaround
		*  as described here:
		*  http://blog.programmingsolution.net/sql-server-2008/tsql/delete-rows-of-a-table-matching-multiple-columns-of-another-table/*/
		String where =
				"(" +
					ElementGeometryTable.Columns.ELEMENT_TYPE +
					LUMP +
					ElementGeometryTable.Columns.ELEMENT_ID +
				")  NOT IN ( " +
					getSelectAllElementsIn(OsmQuestTable.NAME) +
					" UNION " +
					getSelectAllElementsIn(OsmQuestTable.NAME_UNDO) +
				")";

		return db.delete(ElementGeometryTable.NAME, where, null);
	}

	private static final String LUMP = "+'#'+";

	private static String getSelectAllElementsIn(String table)
	{
		return "SELECT " + OsmQuestTable.Columns.ELEMENT_TYPE + LUMP + OsmQuestTable.Columns.ELEMENT_ID +
				" FROM " + table;
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/data/osm/persist/ElementGeometryDao.kt
package de.westnordost.streetcomplete.data.osm.persist

import android.database.Cursor
import android.database.sqlite.SQLiteOpenHelper
import androidx.core.content.contentValuesOf


import javax.inject.Inject

import de.westnordost.streetcomplete.data.osm.ElementGeometry
import de.westnordost.streetcomplete.util.Serializer
import de.westnordost.osmapi.map.data.Element
import de.westnordost.osmapi.map.data.OsmLatLon
import de.westnordost.streetcomplete.data.ObjectRelationalMapping
import de.westnordost.streetcomplete.data.osm.ElementPointGeometry
import de.westnordost.streetcomplete.data.osm.ElementPolygonsGeometry
import de.westnordost.streetcomplete.data.osm.ElementPolylinesGeometry
import de.westnordost.streetcomplete.data.osm.persist.ElementGeometryTable.Columns.ELEMENT_ID
import de.westnordost.streetcomplete.data.osm.persist.ElementGeometryTable.Columns.ELEMENT_TYPE
import de.westnordost.streetcomplete.data.osm.persist.ElementGeometryTable.Columns.GEOMETRY_POLYGONS
import de.westnordost.streetcomplete.data.osm.persist.ElementGeometryTable.Columns.GEOMETRY_POLYLINES
import de.westnordost.streetcomplete.data.osm.persist.ElementGeometryTable.Columns.LATITUDE
import de.westnordost.streetcomplete.data.osm.persist.ElementGeometryTable.Columns.LONGITUDE
import de.westnordost.streetcomplete.data.osm.persist.ElementGeometryTable.NAME
import de.westnordost.streetcomplete.ktx.*

class ElementGeometryDao @Inject constructor(
    private val dbHelper: SQLiteOpenHelper,
    private val mapping: ElementGeometryMapping
) {
    private val db get() = dbHelper.writableDatabase

    fun putAll(entries: Collection<ElementGeometryEntry>) {
        db.transaction {
            for (entry in entries) {
                put(entry)
            }
        }
    }

    fun put(entry: ElementGeometryEntry) {
        val values = contentValuesOf(
            ELEMENT_TYPE to entry.elementType.name,
            ELEMENT_ID to entry.elementId
        ) + mapping.toContentValues(entry.geometry)

        db.replaceOrThrow(NAME, null, values)
    }

    fun get(type: Element.Type, id: Long): ElementGeometry? {
        val where = "$ELEMENT_TYPE = ? AND $ELEMENT_ID = ?"
        val args = arrayOf(type.name, id.toString())

        return db.queryOne(NAME, null, where, args) { mapping.toObject(it) }
    }

    fun delete(type: Element.Type, id: Long) {
        val where = "$ELEMENT_TYPE = ? AND $ELEMENT_ID = ?"
        val args = arrayOf(type.name, id.toString())

        db.delete(NAME, where, args)
    }

    /** Cleans up element geometry entries that belong to elements that are not referenced by any
     * quest anymore.  */
    fun deleteUnreferenced(): Int {
        /* SQLite does not allow selecting multiple columns in a DELETE subquery. Using a workaround
         * as described here:
         * http://blog.programmingsolution.net/sql-server-2008/tsql/delete-rows-of-a-table-matching-multiple-columns-of-another-table/
         */
        val where = "(" + ELEMENT_TYPE + LUMP + ELEMENT_ID + ") NOT IN (" +
            "SELECT " + OsmQuestTable.Columns.ELEMENT_TYPE + LUMP + OsmQuestTable.Columns.ELEMENT_ID + " FROM " + OsmQuestTable.NAME + " " +
            "UNION SELECT " + UndoOsmQuestTable.Columns.ELEMENT_TYPE + LUMP + UndoOsmQuestTable.Columns.ELEMENT_ID + " FROM " + UndoOsmQuestTable.NAME +
            ")"

        return db.delete(NAME, where, null)
    }
}

data class ElementGeometryEntry(
    val elementType: Element.Type,
    val elementId: Long,
    val geometry: ElementGeometry
)

private const val LUMP = "+'#'+"
private typealias PolyLines = ArrayList<ArrayList<OsmLatLon>>

class ElementGeometryMapping @Inject constructor(
    private val serializer: Serializer)
    : ObjectRelationalMapping<ElementGeometry> {

    override fun toContentValues(obj: ElementGeometry) = contentValuesOf(
        LATITUDE to obj.center.latitude,
        LONGITUDE to obj.center.longitude,
        GEOMETRY_POLYGONS to (obj as? ElementPolygonsGeometry)?.let { serializer.toBytes(obj.polygons) },
        GEOMETRY_POLYLINES to (obj as? ElementPolylinesGeometry)?.let { serializer.toBytes(obj.polylines) }
    )

    override fun toObject(cursor: Cursor): ElementGeometry {
        val polylines = cursor.getBlobOrNull(GEOMETRY_POLYLINES)?.let { serializer.toObject<PolyLines>(it) }
        val polygons = cursor.getBlobOrNull(GEOMETRY_POLYGONS)?.let { serializer.toObject<PolyLines>(it) }
        val center = OsmLatLon(cursor.getDouble(LATITUDE), cursor.getDouble(LONGITUDE))

        return when {
            polygons != null ->  ElementPolygonsGeometry(polygons, center)
            polylines != null -> ElementPolylinesGeometry(polylines, center)
            else ->              ElementPointGeometry(center)
        }
    }
}
